File: admin-web/server/src/lib/changeRecord.ts
export const FIELDS = [
  "balance",
  "ramCapacity",
  "callsign",
  "faction",
  "daemons.add",
  "daemons.remove",
  "shards.add",
  "shards.remove",
  "shards.decrypt",
  "counters.breach",
  "counters.alert",
  "counters.blocked",
] as const;
export type Field = (typeof FIELDS)[number];

export const REASONS = [
  "CHARACTER_CREATED",
  "BREACH_ATTEMPT",
  "BREACH_BLOCKED",
  "BREACH_LOOT",
  "BREACH_EDDIES",
  "SHARD_SCAN",
  "SHARD_DECRYPT",
  "TRANSFER_OUT",
  "TRANSFER_IN",
  "TRANSFER_CANCELLED",
  "ITEM_TRANSFER_OUT",
  "ITEM_TRANSFER_IN",
  "ITEM_TRANSFER_CANCELLED",
  "RAM_UPGRADE",
  "ALERT_SENT",
  "ALERT_SUPPRESSED",
  "MASTER_OVERRIDE",
  "CHARACTER_RESET",
] as const;
export type Reason = (typeof REASONS)[number];

/**
 * Запись, как её присылает устройство — сырой JSON из тела запроса, ещё не
 * провалидированный. oldValue/newValue — уже готовая строка (число, короткий
 * текст или JSON.stringify сложного значения), которую сформировало и
 * подписало устройство. Сервер НЕ пересобирает эту строку заново из
 * структурированных данных — берёт байты как есть, иначе подпись не сойдётся
 * из-за расхождений между JSON-сериализаторами Kotlin и JS (порядок ключей,
 * форматирование чисел и т.п.) — та же логика, что у ClaimProtocol.encode.
 */
export interface ChangeRecordInput {
  id: string;
  subjectKeyB64: string;
  seq: number;
  happenedAt: number;
  field: string;
  oldValue: string | null;
  newValue: string | null;
  reason: string;
  sourceRef?: string | null;
  actor: string;
  signature: string;
}

/** Та же запись, уже провалидированная и приведённая к строгим типам полей/причин. */
export interface ChangeRecord extends ChangeRecordInput {
  field: Field;
  reason: Reason;
}

import type { StoredChangeRow } from "../apiTypes.js";

export type { StoredChangeRow };

/**
 * Байты, которые подписывает отправитель — pipe-разделённая строка из полей
 * ровно в том виде, в каком они пришли по проводу (см. комментарий у
 * ChangeRecordInput). Формат сознательно такой же плоский, как у
 * ClaimProtocol.signaturePayload в Android-клиенте, а не JSON — чтобы не
 * зависеть от совпадения сериализаторов на двух платформах.
 */
export function signaturePayload(r: ChangeRecordInput): Buffer {
  const parts = [
    r.id,
    r.subjectKeyB64,
    String(r.seq),
    String(r.happenedAt),
    r.field,
    r.oldValue ?? "",
    r.newValue ?? "",
    r.reason,
    r.sourceRef ?? "",
    r.actor,
  ];
  return Buffer.from(parts.join("|"), "utf8");
}

export function isField(v: string): v is Field {
  return (FIELDS as readonly string[]).includes(v);
}

export function isReason(v: string): v is Reason {
  return (REASONS as readonly string[]).includes(v);
}

File: app/src/main/java/com/megablok10/app/collector/ChangeRecord.kt
package com.megablok10.app.collector

/**
 * Единица обмена с мастерским коллектором (admin-web/server) — см.
 * admin-web/README.md и ТЗ §2.2/§3.1. Зеркало серверного
 * src/lib/changeRecord.ts: те же поля, тот же формат подписи.
 *
 * oldValue/newValue — уже готовая строка (число или JSON.stringify сложного
 * значения), не структура: сервер сверяет подпись побайтово по тому, что
 * реально легло в HTTP-тело, не пересобирая её заново — так же, как
 * ClaimProtocol.signaturePayload для заявок на слоты. Если здесь положить
 * структуру и сериализовать её отдельно на каждом конце, подписи разойдутся
 * из-за разницы JSON-сериализаторов Kotlin/JS.
 */
data class ChangeRecord(
    val id: String,
    val subjectKeyB64: String,
    val seq: Long,
    val happenedAt: Long,
    val field: String,
    val oldValue: String?,
    val newValue: String?,
    val reason: String,
    val sourceRef: String?,
    val actor: String,
    val signature: String,
)

object ChangeField {
    const val BALANCE = "balance"
    const val RAM_CAPACITY = "ramCapacity"
    const val CALLSIGN = "callsign"
    const val FACTION = "faction"
    const val DAEMONS_ADD = "daemons.add"
    const val DAEMONS_REMOVE = "daemons.remove"
    const val SHARDS_ADD = "shards.add"
    const val SHARDS_REMOVE = "shards.remove"
    const val SHARDS_DECRYPT = "shards.decrypt"
    const val COUNTERS_BREACH = "counters.breach"
    const val COUNTERS_ALERT = "counters.alert"
    const val COUNTERS_BLOCKED = "counters.blocked"

    /** Только от мастера (дашборд → устройство): текст объявления. Устройство такие записи не создаёт. */
    const val ANNOUNCEMENT = "announcement"
}

object ChangeReason {
    const val CHARACTER_CREATED = "CHARACTER_CREATED"
    const val BREACH_ATTEMPT = "BREACH_ATTEMPT"
    const val BREACH_BLOCKED = "BREACH_BLOCKED"
    const val BREACH_LOOT = "BREACH_LOOT"
    const val BREACH_EDDIES = "BREACH_EDDIES"
    const val SHARD_SCAN = "SHARD_SCAN"
    const val SHARD_DECRYPT = "SHARD_DECRYPT"
    const val TRANSFER_OUT = "TRANSFER_OUT"
    const val TRANSFER_IN = "TRANSFER_IN"
    const val TRANSFER_CANCELLED = "TRANSFER_CANCELLED"
    const val ITEM_TRANSFER_OUT = "ITEM_TRANSFER_OUT"
    const val ITEM_TRANSFER_IN = "ITEM_TRANSFER_IN"
    const val ITEM_TRANSFER_CANCELLED = "ITEM_TRANSFER_CANCELLED"
    const val RAM_UPGRADE = "RAM_UPGRADE"
    const val ALERT_SENT = "ALERT_SENT"
    const val ALERT_SUPPRESSED = "ALERT_SUPPRESSED"
    const val MASTER_OVERRIDE = "MASTER_OVERRIDE"
    const val CHARACTER_RESET = "CHARACTER_RESET"
}

/** Байты, которые подписываются — тот же плоский pipe-формат, что signaturePayload на сервере (admin-web/server/src/lib/changeRecord.ts). */
fun ChangeRecord.signaturePayload(): ByteArray {
    val parts = listOf(
        id, subjectKeyB64, seq.toString(), happenedAt.toString(), field,
        oldValue ?: "", newValue ?: "", reason, sourceRef ?: "", actor,
    )
    return parts.joinToString("|").toByteArray(Charsets.UTF_8)
}
